### DISCUSSION TITLE
"Officially" Expose image content or path in content collection schema definition

### DISCUSSION THREAD

**Discussion by aripalo** [withastro/roadmap#823] [Proposal]
# Summary

Relates to discussion in this issue: https://github.com/withastro/astro/issues/9852

When defining content collection schemas, expose either image content (Buffer) or path (string) through `ctx.image()`. 

# Background & Motivation

This would be useful for example with generating placeholder images (like [blurhash](https://blurha.sh/) / [thumbhash](https://evanw.github.io/thumbhash/) / etc) _centrally_, avoiding the need to implement such logic in various places around the codebase (in layouts/components).

Once available through the "native" content collection schema, one could use the value in any component/layout or even in API routes (that for example might generate OpenGraph images).

## Workarounds
- As described in the example code in the [original issue](https://github.com/withastro/astro/issues/9852#issue-2104184880), of course I can access the filesystem path by checking the existance of `ASTRO_ASSET`, but – as stated by [Princesseuh](https://github.com/Princesseuh) in [this comment](https://github.com/withastro/astro/issues/9852#issuecomment-1913654748) – that is internal "API" and not to be relied on.
- One way of accessing a image frontmatter definition could be using remark plugins (see for example ["add reading time" recipe](https://docs.astro.build/en/recipes/reading-time/), but even if it was possibly, the custom frontmatter added via remark & co tools is not (type)safely available by default when querying collections with `getCollection()`.

# Goals

- Expose a way to access either
   1. the image content (for example Buffer) 
   2. or image path (this could be filepath or remote path)
   
   Option `i` would be preferred as it wouldn't introduce duplicate fetching of the image content.

- This "API" (or some field in the [return type of `ctx.image()`](https://github.com/withastro/astro/blob/70fdf1a5c660057152c1ca111dcc89ceda5c8840/packages/astro/content-types.template.d.ts#L23) could be marked with some warnings etc like "avoid this", but at least make it "available", so one could rely on it if needed.

# Example

https://stackblitz.com/edit/github-avs1k5?file=src%2Fcontent%2Fconfig.ts

**Comment by aripalo** [comment]
And just to clarify, there are workarounds, but I just think this would be useful feature to have "out-of-the-box".

**Comment by Princesseuh** [comment]
This is a bit challenging to support because of SSR, where you don't have access to the original paths anymore, and the final path of the images might be somewhere else than the expected position. I think it'd still be possible, though.

Buffer is similar, we can't include the image data in the runtime bundle (because it'd be absurdly large), so it'd need to be lazy, but in SSR you don't necessarily have access to filesystem APIs, so everything is annoying.

SSR makes my life complicated 😅

**Reply by tylermercer** [reply]
Could the source filesystem path be exposed with a naming that makes it clear it's the path on the filesystem at build time? It wouldn't be directly useful in SSR contexts for the reasons you described, but this would at least open up build-time use cases (like build-time rendered image placeholders and OG images, both of which I have implemented in https://evelynescobar.art using a manual helper function to reconstruct filesystem asset paths, which is a bit hacky).
